### ISSUE TITLE
Problems with graceful-fs/polyfills.js

### ISSUE THREAD

**Issue by raphael10-collab** [isaacs/node-graceful-fs#196]
I'm encountering this problem:

I discovered that once I activate this line:

 const stateRecord: App = loadSettings()

in /src/store/modules/app.ts  

I get this error: "Cannot read property 'match' of undefined"

![image](https://user-images.githubusercontent.com/70313067/92303609-ccd65780-ef76-11ea-978d-2e5413aa31f8.png)

This is /src/store/modules/app.ts :  

    import Vue from 'vue'
    import { loadSettings, setSettings } from '@/services/electron-services/setting/setting'

    const TOGGLE_THEME = 'TOGGLE_THEME'
    const TOGGLE_LANG = 'TOGGLE_LANG'

    const stateRecord: App = loadSettings()

    const app = {
      state: {
        currentTheme: stateRecord.currentTheme || 'light',
      },
      mutations: {
        [TOGGLE_THEME](state: App, currentTheme: Theme) {
          state.currentTheme = currentTheme
        },
      },
      actions: {
        TOGGLE_THEME({ commit }: any, payload: App) {
          setSettings('settings.currentTheme', payload.currentTheme)
          commit(TOGGLE_THEME, payload.currentTheme)
        },
      },
    }

    export default app

this is src/services/electron-services/setting/setting :

    import db from '../database/index'

    export const loadSettings = (): App => {
      return db.get<App>('settings')
    }

    export const setSettings = (key: string, value: string | boolean | number): string | boolean 
    | number => {
      return db.set<string | boolean | number>(key, value)
    }

    export default {}

and this is electron-services/database/index.ts :

    import Lowdb from 'lowdb'
    import FileSync from 'lowdb/adapters/FileSync'
    import path from 'path'
    import fs from 'fs-extra'
    import LodashID from 'lodash-id'
    import { app, remote } from 'electron'

    interface Schema {
      windowSize: {
        height: number
        width: number
      }
      settings: {
        currentLang: string
        currentTheme: string
      }
    }

    const isRenderer: boolean = process.type === 'renderer'
    // Render process use remote app
    const APP: Electron.App = isRenderer ? remote.app : app

    const STORE_PATH: string = APP.getPath('userData')

    // In production mode, during the first open application
    // APP.getPath('userData') gets the path nested and the datastore.js is loaded.
    // if it doesn't exist, create it.
    if (!isRenderer) {
      if (!fs.pathExistsSync(STORE_PATH)) {
        fs.mkdirpSync(STORE_PATH)
      }
    }

    class DB {
      private db: Lowdb.LowdbSync<Schema>
      public constructor() {
        const adapter: Lowdb.AdapterSync<Schema> = new FileSync<Schema>(path.join(STORE_PATH, 
    '/db.json'))
        this.db = Lowdb(adapter)
        // Use lodash-id must use insert methods
        this.db._.mixin(LodashID)
        if (!this.db.has('windowSize').value()) {
          this.db
            .set('windowSize', {
              width: 1025,
              height: 749,
            })
            .write()
        }
        if (!this.db.has('settings').value()) {
          this.db
            .set('settings', {
              currentLang: 'en',
              currentTheme: 'light',
            })
            .write()
        }
        // Purple to Night
        if (this.db.get('settings.currentTheme').value() === 'purple') {
          this.db.set('settings.currentTheme', 'night').write()
        }
        if (!this.db.has('settings.currentLang')) {
          this.db.set('settings.currentLang', 'en').write()
        }
      }
      // read() is to keep the data of the main process and the rendering process up to date.
      public read() {
        return this.db.read()
      }
      public get<T>(key: string): T {
        return this.read().get(key).value()
        //return this.db.get(key).value()
      }
      public find<T>(key: string, id: string): T {
        const data: $TSFixed = this.read().get(key)
        return data.find({ id }).value()
      }
      public set<T>(key: string, value: T): T {
        //return this.read().set(key, value).write()
        return this.db.set(key, value).write()
      }
      public insert<T>(key: string, value: T): T {
        const data: $TSFixed = this.read().get(key)
        return data.insert(value).write()
      }
      public update<T>(key: string, id: string, value: T): T {
        const data: $TSFixed = this.read().get(key)
        return data.find({ id }).assign(value).write()
      }
      public remove<T>(key: string, id: string): T {
        const data: $TSFixed = this.read().get(key)
        return data.removeById(id).write()
      }
      public filter<T, K>(key: string, query: K): T {
        const data: $TSFixed = this.read().get(key)
        return data.filter(query).value()
      }
      public has(key: string): boolean {
        return this.read().has(key).value()
      }
    }

The operation setSettings has the same problem:
once I only activate setSettings in src/store/modules/app.ts  I get the same error:

    import Vue from 'vue'
    //import { loadSettings, setSettings } from '@/services/electron-services/setting/setting'
    import { setSettings } from '@/services/electron-services/setting/setting'

    const TOGGLE_THEME = 'TOGGLE_THEME'
    const TOGGLE_LANG = 'TOGGLE_LANG'

    //const stateRecord: App = loadSettings()

    const app = {
      state: {
        //currentTheme: stateRecord.currentTheme || 'dark',
        //currentLang: stateRecord.currentLang || 'en',
        //currentTheme: 'light',
      },
      mutations: {
        //[TOGGLE_THEME](state: App, currentTheme: Theme) {
        //  state.currentTheme = currentTheme
        //},
        //[TOGGLE_LANG](state: App, currentLang: Language) {
        //  state.currentLang = currentLang
        //},
      },
      actions: {
        TOGGLE_THEME({ commit }: any, payload: App) {
          setSettings('settings.currentTheme', payload.currentTheme)
          commit(TOGGLE_THEME, payload.currentTheme)
        },
        //TOGGLE_LANG({ commit }: any, payload: App) {
        //  setSettings('settings.currentLang', payload.currentLang)
        //  commit(TOGGLE_LANG, payload.currentLang)
        //},
      },
    }

    export default app

src/services/electron-services/setting/setting.ts : 

    import db from '../database/index'

    //export const loadSettings = (): App => {
    //  return db.get<App>('settings')
    //}

    export const setSettings = (key: string, value: string | boolean | number): string | boolean | number => {
      return db.set<string | boolean | number>(key, value)
    }

    export default {}

![image](https://user-images.githubusercontent.com/70313067/92303641-2c346780-ef77-11ea-8dd0-40a47dde97f1.png)

Also  modifying

    public constructor() {
        // const adapter: Lowdb.AdapterSync<Schema> = new FileSync<Schema>(path.join(STORE_PATH, '/db.json'))
        const adapter= new FileSync('db.json')

still get this error

![image](https://user-images.githubusercontent.com/70313067/92303654-5dad3300-ef77-11ea-9f68-4ea01fe61891.png)

This is the line in fileSync.js which seems to be involved:

![image](https://user-images.githubusercontent.com/70313067/92303661-6d2c7c00-ef77-11ea-9810-15ba08ad7322.png)

It seems that the line involved in node_modules/graceful-fs/polyfills.js is the following: 

     process.version.match(/^v0\.6\.[0-2]|^v0\.5\./))

in this function:

    function patch (fs) {
      // (re-)implement some things that are known busted or missing.

      // lchmod, broken prior to 0.6.2
      // back-port the fix here.
      if (constants.hasOwnProperty('O_SYMLINK') &&
        process.version.match(/^v0\.6\.[0-2]|^v0\.5\./)) {   // <------------------------------------------------------------------------------------
        patchLchmod(fs)
      }

vue info:

    Environment Info:

      System:
        OS: Linux 5.4 Ubuntu 18.04.5 LTS (Bionic Beaver)
        CPU: (8) x64 Intel(R) Core(TM) i7-4790K CPU @ 4.00GHz
      Binaries:
        Node: 14.5.0 - ~/.nvm/versions/node/v14.5.0/bin/node
        Yarn: 1.22.5 - /usr/bin/yarn
        npm: 6.14.5 - ~/.nvm/versions/node/v14.5.0/bin/npm
      Browsers:
        Chrome: 85.0.4183.83
        Firefox: 80.0
      npmPackages:
        @vue/babel-helper-vue-jsx-merge-props:  1.0.0 
        @vue/babel-plugin-transform-vue-jsx:  1.1.2 
        @vue/babel-preset-app:  4.4.6 
        @vue/babel-preset-jsx:  1.1.2 
        @vue/babel-sugar-functional-vue:  1.1.2 
        @vue/babel-sugar-inject-h:  1.1.2 
        @vue/babel-sugar-v-model:  1.1.2 
        @vue/babel-sugar-v-on:  1.1.2 
        @vue/cli-overlay:  4.4.6 
        @vue/cli-plugin-babel: ~4.4.0 => 4.4.6 
        @vue/cli-plugin-e2e-cypress: ~4.4.0 => 4.4.6 
        @vue/cli-plugin-router: ~4.4.0 => 4.4.6 
        @vue/cli-plugin-typescript: ~4.4.0 => 4.4.6 
        @vue/cli-plugin-unit-mocha: ~4.4.0 => 4.4.6 
        @vue/cli-plugin-vuex: ~4.4.0 => 4.4.6 
        @vue/cli-service: ~4.4.0 => 4.4.6 
        @vue/cli-shared-utils:  4.4.6 
        @vue/component-compiler-utils:  3.2.0 
        @vue/preload-webpack-plugin:  1.1.2 
        @vue/test-utils: ^1.0.3 => 1.0.3 
        @vue/web-component-wrapper:  1.2.0 
        babel-helper-vue-jsx-merge-props:  2.0.3 
        typescript: ^3.9.7 => 3.9.7 
        vue: ^2.6.11 => 2.6.11 
        vue-class-component: ^7.2.5 => 7.2.5 
        vue-cli-plugin-electron-builder: ~2.0.0-rc.4 => 2.0.0-rc.4 
        vue-hot-reload-api:  2.3.4 
        vue-i18n: ^8.20.0 => 8.20.0 
        vue-loader:  15.9.3 
        vue-property-decorator: ^9.0.0 => 9.0.0 
        vue-router: ^3.2.0 => 3.3.4 
        vue-style-loader:  4.1.2 
        vue-template-compiler: ^2.6.11 => 2.6.11 
        vue-template-es2015-compiler:  1.9.1 
        vuex: ^3.5.1 => 3.5.1 
        vuex-class: ^0.3.2 => 0.3.2 
      npmGlobalPackages:
        @vue/cli: 4.4.6

      electron version: 10.0.0

This is vue.config.js :   

    const WorkerPlugin = require('worker-plugin')

    module.exports = {
      // options...
      publicPath: '',
      pluginOptions: {
        electronBuilder: {
          // Prevent bundling of certain imported packages and instead retrieve these external 
    dependencies at runtime.
          // In order to connect to websocket.
          // https://github.com/nklayman/vue-cli-plugin-electron-builder/issues
    /652#issuecomment-583764345
          externals: ['ggc', 'tesseract.js'],
          builderOptions: {
            productName: 'GGC',
            win: {
              icon: './public/app.ico'
            },
            mac: {
              icon: './public/icons/Icon.icns',
              target: [
                'pkg',
                'dmg',
                'zip',
              ],
            },
            linux: {
              icon: './public/app.png'
            }
          },
          // https://nklayman.github.io/vue-cli-plugin-electron-builder/guide
    /configuration.html#webpack-configuration
          chainWebpackRendererProcess: (config) => {
            // Chain webpack config for electron renderer process only
            // The following example will set IS_ELECTRON to true in your app
            config.plugin('define').tap((args) => {
              args[0]['IS_ELECTRON'] = true
              return args
            })
          },
          mainProcessFile: 'src/background.ts',
          // https://nklayman.github.io/vue-cli-plugin-electron-builder/guide
    /configuration.html#typescript-options
          disableMainProcessTypescript: false, // Manually disable typescript plugin for main 
    process. Enable if you want to use regular js for the main process (src/background.js by 
    default)
          mainProcessTypeChecking: false, // Manually enable type checking during webpck 
    bundling for background file.
          // https://nklayman.github.io/vue-cli-plugin-electron-builder/guide
    /guide.html#preload-files
          preload: 'src/preload.js',
          // https://nklayman.github.io/vue-cli-plugin-electron-builder/guide
    /security.html#node-integration
          nodeIntegration: false
        },
        // https://nklayman.github.io/vue-cli-plugin-electron-builder/guide/guide.html#web-
    workers
        configureWebpack: {
          plugins: [new WorkerPlugin()]
        }
      }
    }

And this is webpack.config.js :  

    module.exports = {
      entry: './src/background.ts',
      target: 'node',
      output: {
        path: path.join(__dirname, 'build'),
        filename: 'background.js'
      },
      // https://github.com/GoogleChromeLabs/worker-plugin
      plugins: [
        new WorkerPlugin()
      ],
      // https://webpack.js.org/plugins/define-plugin/
      new webpack.DefinePlugin({
        'process.env.NODE_ENV': JSON.stringify(process.env.NODE_ENV)
      }),
      //https://vuetifyjs.com/en/getting-started/quick-start/#webpack-install
      rules: [
        {
          test: /\.s(c|a)ss$/,
          use: [
            'vue-style-loader',
            'css-loader',
            {
              loader: 'sass-loader',
              // Requires sass-loader@^7.0.0
              options: {
                implementation: require('sass'),
                fiber: require('fibers'),
                indentedSyntax: true // optional
              },
              // Requires sass-loader@^8.0.0
              options: {
                implementation: require('sass'),
                sassOptions: {
                  fiber: require('fibers'),
                  indentedSyntax: true // optional
                },
              },
            },
          ],
        },
      ],
    }

What could be the cause of this "Cannot read property 'match' of undefined"? What's wrong with this function in class DB ?:

    public get<T>(key: string): T {
      return this.read().get(key).value()
      //return this.db.get(key).value()
    }

      electron version: 10.0.0

This is vue.config.js :   

    const WorkerPlugin = require('worker-plugin')

    module.exports = {
      // options...
      publicPath: '',
      pluginOptions: {
        electronBuilder: {
          // Prevent bundling of certain imported packages and instead retrieve these external 
    dependencies at runtime.
          // In order to connect to websocket.
          // https://github.com/nklayman/vue-cli-plugin-electron-builder/issues
    /652#issuecomment-583764345
          externals: ['ggc', 'tesseract.js'],
          builderOptions: {
            productName: 'GGC',
            win: {
              icon: './public/app.ico'
            },
            mac: {
              icon: './public/icons/Icon.icns',
              target: [
                'pkg',
                'dmg',
                'zip',
              ],
            },
            linux: {
              icon: './public/app.png'
            }
          },
          // https://nklayman.github.io/vue-cli-plugin-electron-builder/guide
    /configuration.html#webpack-configuration
          chainWebpackRendererProcess: (config) => {
            // Chain webpack config for electron renderer process only
            // The following example will set IS_ELECTRON to true in your app
            config.plugin('define').tap((args) => {
              args[0]['IS_ELECTRON'] = true
              return args
            })
          },
          mainProcessFile: 'src/background.ts',
          // https://nklayman.github.io/vue-cli-plugin-electron-builder/guide
    /configuration.html#typescript-options
          disableMainProcessTypescript: false, // Manually disable typescript plugin for main 
    process. Enable if you want to use regular js for the main process (src/background.js by 
    default)
          mainProcessTypeChecking: false, // Manually enable type checking during webpck 
    bundling for background file.
          // https://nklayman.github.io/vue-cli-plugin-electron-builder/guide
    /guide.html#preload-files
          preload: 'src/preload.js',
          // https://nklayman.github.io/vue-cli-plugin-electron-builder/guide
    /security.html#node-integration
          nodeIntegration: false
        },
        // https://nklayman.github.io/vue-cli-plugin-electron-builder/guide/guide.html#web-
    workers
        configureWebpack: {
          plugins: [new WorkerPlugin()]
        }
      }
    }

You can find the GitHub repo here: https://github.com/raphael10-collab/ElectronVueTypeScriptScaffolding

Looking forward to your kind help

**Comment by truefusion** [comment]
Here is a workaround I came across:
```js
const { remote } = window.require('electron');
const fs = remote.require('fs-extra');
```

**Comment by raphael10-collab** [comment]
Hi @truefusion 
thanks for answering.

I do not understand what I should do.

In src/services/electron-services/database/index.ts I tried to substitute 

    //import fs from 'fs-extra'

with 

    import pathExistsSync from 'fs-extra/pathExistsSync'
    import mkdirpSync from 'fs-extra/mkdirpSync'

    if (!isRenderer) {
      //if (!fs.pathExistsSync(STORE_PATH)) {
        //fs.mkdirpSync(STORE_PATH)
      if (!pathExistsSync(STORE_PATH)) {
        mkdirpSync(STORE_PATH)
      }
    }

but got "Cannot find module 'fs-extra/pathExistsSync' or its corresponding type declarations" 
and "Cannot find module 'fs-extra/mkdirpSync' or its corresponding type declarations."

**Comment by truefusion** [comment]
I couldn't provide you with relevant code because I didn't know what exactly you were trying to do. But if you were trying to use the `fs-extra` module in an electron environment, then the code I provided should be enough. The code you responded with did not make use of the code I provided. I will assume you know how to program in JavaScript and about the many different ways to import or require modules, so I won't go into details for that. But if you are trying to use `fs-extra` in an electron environment, then you must use the code I provided in exactly the way I provided it.

In other words, this code
```js
import fs from 'fs-extra'
```

Becomes this code
```js
const { remote } = window.require('electron');
const fs = remote.require('fs-extra');
```

Then you can start using `fs` anywhere. For example,
```js
const { remote } = window.require('electron');
const fs = remote.require('fs-extra');

if (!isRenderer) {
  if (!fs.pathExistsSync(STORE_PATH)) {
    fs.mkdirpSync(STORE_PATH)
  }
}
```

I don't know where `isRenderer` came from, but I don't need to know.

**Comment by raphael10-collab** [comment]
@truefusion where does window come from?

**Comment by truefusion** [comment]
`window` is a pre-defined, global variable. Therefore it comes from the environment, in this case it would be from electron's environment. If you were running your script in a node environment, I would not expect `window` to be defined.

**Comment by raphael10-collab** [comment]
@truefusion 
I tried to put

    //import fs from 'fs-extra'

    const { BrowserWindow } = require('electron').remote
    let window = new BrowserWindow()
    const { remote } = window.require('electron');
    const fs = remote.require('fs-extra');
    import LodashID from 'lodash-id'
    //import { app, remote } from 'electron'
    import { app } from 'electron'

But I got this error:

    A JavaScript error occurred in the main process
    Uncaught Exception:
    TypeError: Cannot destructure property 'BrowserWindow' of '__webpack_require__(...).remote' as it is undefined.
        at eval (webpack:///./src/services/electron-services/database/index.ts?:16:9)
        at Module../src/services/electron-services/database/index.ts (/home/marco/webMatters/electronMatters/GGC-Electron
    /dist_electron/index.js:1984:1)
        at __webpack_require__ (/home/marco/webMatters/electronMatters/GGC-Electron/dist_electron/index.js:20:30)
        at eval (webpack:///./src/background.ts?:7:100)
        at Module../src/background.ts (/home/marco/webMatters/electronMatters/GGC-Electron/dist_electron/index.js:1960:1)
        at __webpack_require__ (/home/marco/webMatters/electronMatters/GGC-Electron/dist_electron/index.js:20:30)
        at eval (webpack:///multi_./src/background.ts?:1:18)
        at Object.0 (/home/marco/webMatters/electronMatters/GGC-Electron/dist_electron/index.js:2007:1)
        at __webpack_require__ (/home/marco/webMatters/electronMatters/GGC-Electron/dist_electron/index.js:20:30)
        at /home/marco/webMatters/electronMatters/GGC-Electron/dist_electron/index.js:84:18

actually window, aka win, is already defined in src/background.ts :   

    import { app, protocol, BrowserWindow, ipcMain, shell, Menu } from 'electron'

    async function createWindow() {
      win = new BrowserWindow({

But window has to be defined in src/services/electron-services/database/index.ts as well in order to be able to do 
const { remote } = window.require('electron')

**Comment by truefusion** [comment]
The moment you redefine `window` is when you break its original value. I'm not sure what your programming background is, but there is a bigger underlying issue here. It is best if you take this time to come to a deeper understanding of JavaScript and its environments before continuing to any further on your project.

Here may be a good place to start: https://developer.mozilla.org/en-US/docs/Web/JavaScript

**Comment by raphael10-collab** [comment]
@truefusion thank for your kind suggestion.
I would be grateful if you would have a look at the file where you suggest to modify how to import fs-extra : https://github.com/raphael10-collab/ElectronVueTypeScriptScaffolding/blob/master/src/services/electron-services/database/index.ts and tell me how to use window, which cannot be redefined, otherwise its original value will be broken, there

**Comment by raphael10-collab** [comment]
@truefusion thank you for your pull request
Doing :  

    const { app, remote } = window.require('electron')
    import Lowdb from 'lowdb'
    import FileSync from 'lowdb/adapters/FileSync'
    import path from 'path'
    const fs = remote.require('fs-extra')

I get : "ReferenceError: window is not defined".
So I have to understand how to define window in src/services/electron-services/database/index.ts

**Comment by truefusion** [comment]
How are you compiling your code? And how are you running your app?

**Comment by raphael10-collab** [comment]
This is the complete output:

    (base) marco@pc01:~/webMatters/electronMatters/ElectronVueTypeScriptScaffolding$ yarn electron:serve
    yarn run v1.22.5
    warning ../package.json: No license field
    $ vue-cli-service electron:serve
     INFO  Starting development server...
    Starting type checking service...
    Using 1 worker with 2048MB memory limit
    98% after emitting CopyPlugin

     DONE  Compiled successfully in 1725ms                                                                                                                                         
    10:26:28 PM

    Type checking in progress...

      App running at:
      - Local:   http://localhost:8080 
      - Network: http://192.168.1.7:8080

      Note that the development build is not optimized.
      To create a production build, run yarn build.

    ⠏  Bundling main process...

     DONE  Compiled successfully in 855ms                                                                                                                                       
    10:26:29 PM

      File                      Size                                                                       Gzipped

      dist_electron/index.js    1366.33 KiB                                                                274.09 KiB

     Images and other types of assets omitted.



      DONE  Compiled successfully in 11ms       
     10:26:29 PM


      File                        Size                                                                      Gzipped

      dist_electron/preload.js    76.96 KiB                                                                 5.19 KiB

      Images and other types of assets omitted.

     INFO  Launching Electron...
    App threw an error during load
    ReferenceError: window is not defined
        at eval (webpack:///./src/services/electron-services/database/index.ts?:8:25)
        at Module../src/services/electron-services/database/index.ts (/home/marco/webMatters/electronMatters
    /ElectronVueTypeScriptScaffolding/dist_electron/index.js:1894:1)
        at __webpack_require__ (/home/marco/webMatters/electronMatters/ElectronVueTypeScriptScaffolding/dist_electron
    /index.js:20:30)
        at eval (webpack:///./src/background.ts?:7:100)
        at Module../src/background.ts (/home/marco/webMatters/electronMatters/ElectronVueTypeScriptScaffolding/dist_electron
    /index.js:1870:1)
        at __webpack_require__ (/home/marco/webMatters/electronMatters/ElectronVueTypeScriptScaffolding/dist_electron
    /index.js:20:30)
        at eval (webpack:///multi_./src/background.ts?:1:18)
        at Object.0 (/home/marco/webMatters/electronMatters/ElectronVueTypeScriptScaffolding/dist_electron/index.js:1917:1)
    at __webpack_require__ (/home/marco/webMatters/electronMatters/ElectronVueTypeScriptScaffolding/dist_electron
    /index.js:20:30)
        at /home/marco/webMatters/electronMatters/ElectronVueTypeScriptScaffolding/dist_electron/index.js:84:18
    A JavaScript error occurred in the main process
    Uncaught Exception:
    ReferenceError: window is not defined
        at eval (webpack:///./src/services/electron-services/database/index.ts?:8:25)
        at Module../src/services/electron-services/database/index.ts (/home/marco/webMatters/electronMatters
    /ElectronVueTypeScriptScaffolding/dist_electron/index.js:1894:1)
        at __webpack_require__ (/home/marco/webMatters/electronMatters/ElectronVueTypeScriptScaffolding/dist_electron
    /index.js:20:30)
        at eval (webpack:///./src/background.ts?:7:100)
        at Module../src/background.ts (/home/marco/webMatters/electronMatters/ElectronVueTypeScriptScaffolding/dist_electron
    /index.js:1870:1)
        at __webpack_require__ (/home/marco/webMatters/electronMatters/ElectronVueTypeScriptScaffolding/dist_electron
    /index.js:20:30)
        at eval (webpack:///multi_./src/background.ts?:1:18)
        at Object.0 (/home/marco/webMatters/electronMatters/ElectronVueTypeScriptScaffolding/dist_electron/index.js:1917:1)
        at __webpack_require__ (/home/marco/webMatters/electronMatters/ElectronVueTypeScriptScaffolding/dist_electron
    /index.js:20:30)
        at /home/marco/webMatters/electronMatters/ElectronVueTypeScriptScaffolding/dist_electron/index.js:84:18
    No type errors found
    Version: typescript 3.9.7
    Time: 3989ms

**Comment by truefusion** [comment]
Your `dist_electron/preload.js` script is not designed to properly launch electron.

**Comment by raphael10-collab** [comment]
@truefusion 

This is the preload.js I'm using, based on the suggestions here: https://github.com/electron/electron/issues/9920#issuecomment-575839738


    const {
        contextBridge,
        ipcRenderer
    } = require("electron");

    // Expose protected methods that allow the renderer process to use
    // the ipcRenderer without exposing the entire object
    contextBridge.exposeInMainWorld(
        "api", {
            send: (channel, data) => {
                // whitelist channels
                let validChannels = ["toMain"];
                if (validChannels.includes(channel)) {
                    ipcRenderer.send(channel, data);
                }
            },
            receive: (channel, func) => {
                let validChannels = ["fromMain"];
                if (validChannels.includes(channel)) {
                    // Deliberately strip event as it includes `sender` 
                    ipcRenderer.on(channel, (event, ...args) => func(...args));
                }
            }
        }
    );

How should I modify preload.js in order to properly launch electron?

**Comment by truefusion** [comment]
This is as far as I will help you on the matter. https://www.electronjs.org/docs/tutorial/quick-start

For anything else, you'll have to become more acquainted with web technologies.

**Comment by raphael10-collab** [comment]
Thank you very much for tour kind valuable contribution

Il gio 17 set 2020, 21:11 truefusion <notifications@github.com> ha scritto:

**Comment by nmschorr** [comment]
I have the same problem:  

Uncaught TypeError: can't access property "match", e.version is undefined
    c polyfills.js:30
    v graceful-fs.js:104
    0da7 graceful-fs.js:96
    0da7 chunk-vendors.331b15a4.js:1
    Webpack 14
polyfills.js:30:16
    c polyfills.js:30
    v graceful-fs.js:104
    0da7 graceful-fs.js:96
    0da7 chunk-vendors.331b15a4.js:1
    Webpack 14
      
I've tried everything under the sun, and don't think it's my npm installation.

Let me know if you would like more info.  Meanwhile I'm going to try commenting out:

   if (constants.hasOwnProperty('O_SYMLINK') &&
      process.version.match(/^v0\.6\.[0-2]|^v0\.5\./)) {
    patchLchmod(fs)
  }

which seems to be the problem area so far.

**Comment by schoudhary21** [comment]
i faced the same issue and fixed it by downgrading to graceful-fs "4.2.4" version; force it using 'npm install graceful-fs@4.2.4'
the recent versions breaking for me

**Comment by Clockwerk300** [comment]
You need to make sure that dart-sass is not loaded, because this sass cannot run in the browser yet.

I had to remove "implementation: require('sass')," in my vue.config.js and then it worked.

**Comment by 1c7** [comment]
## webpack config, try this, may works
```javascript
plugins: [
  // https://webpack.js.org/plugins/environment-plugin/
  new webpack.DefinePlugin({
    'process.version': process.version,
  }), 
],
```

**Comment by eumenhofer** [comment]
Running into this as well with react on the latest version. Workarounds aren't really an option for my project.

**Comment by calix-cui** [comment]
This problem is most likely caused by Webpack5. Webpack5 no longer automatically populates the nodejs core module.
https://webpack.js.org/configuration/resolve/#resolvefallback
![image](https://user-images.githubusercontent.com/91196517/228416265-2c2b9b36-d56b-45bc-966e-5e7e3f3b1031.png)
